### PR TITLE
Thread subheaders and parent navigation

### PR DESCRIPTION
**Pull request by chiragsalian**
### Details
Adding subheaders and parent navigation for child reports. The latest on where we landed for this is mentioned [here](https://expensify.slack.com/archives/C02HWMSMZEC/p1687292113508279?thread_ts=1686946331.451439&cid=C02HWMSMZEC)

i.e.,
- If the report discussion is tied to a workspace, the name logic takes the form: `From <parent report name> in <workspace name>`
    - Because the workspace chat is the name of the workspace for members, that header takes the form: `From <parent report name>`
- If the report discussion is not tied to a workspace, the name logic takes the form: `From <parent report name>`

### Fixed Issues
$ https://github.com/Expensify/App/issues/20844
PROPOSAL: GH_LINK_ISSUE(COMMENT)

#### Regular DMs
1. Start a DM between userA and userB.
2. Create a thread. Have both users visit the thread.
3. Confirm userA sees the subheader `From <userB full name>`. (its basically `From <parent report name>`)
4. Confirm userB sees the subheader `From <userA full name>`
5. Click on the subheader and confirm you are navigated up one level to parent. 
image

#### Group DMs
1. Start a DM between userA, userB and userC.
2. Create a thread. Have userA visit the thread.
3. Confirm userA sees the subheader `From <userB>, <userC>` (i.e., `From <parent report name>`)
4. Click on the subheader and confirm you are navigated up one level to parent. 
image

### Task DM
1. Start a DM between userA and userB.
2. Create a task. Have both users visit the task. (you might have to refresh or exit and revisit this task report because at the time of writing optimistic data and onyx data are not properly received so a refresh or revisit should fetch the latest data)
3. Confirm userA sees the subheader `From <userB full name>`. (its basically `From <parent report name>`)
4. Confirm userB sees the subheader `From <userA full name>`
5. Click on the subheader and confirm you are navigated up one level to parent. 
image

### IOUs
1. Start a DM between userA and userB.
2. Click on the plus actions icon next to "write something.." in the compose message box and select "request money".
3. Click on the expense request preview. 
4. In the expense report, confirm the subheader says `From <userA full name>`
image

### IOUs children
1. Continuing from previous. Click on the specific expense in the expense report.
2. Confirm you see the single expense request view. Confirm the subheader says `From <parent IOU report name>` 
The parent IOU report can be seen in the LHN. In the below screenshot the parent IOU report name is "Marc Glasser owes $20.00" and so the child has that as its subheader.
image
3. Click on the subheader and confirm you are navigated up one level to parent. 
4. Create another thread in the expense report, Or a few nested thread. Confirm all thread subheaders show `From <parent IOU report name>`.
image

### Policy room threads
1. Create a public room. Have userA and userB join the public room.
6. Create a thread. Have both users visit the thread.
7. Confirm userA and userB sees the subheader `From <public room name> in <workspace name>`
8. Create another thread and confirm the subheader is the same.
9. Click on the subheader and confirm you are navigated up one level to parent.
image

### Policy room tasks
1. Create a public room. Have userA and userB join the public room.
5. Create a task and assign it to userB. The task report most likely should automatically open, if not click into it (if you do not see a subheader you need to refresh or revisit the task to fetch the latest data)
6. Confirm userA sees the subheader `From <public room name> in <workspace name>`.
image

### Expense Chats
1. Create a workspace. As admin userA, invite userB to the workspace.
2. You should see a workspace chat between the two like so (view for the admin is below, for a member it will look slightly different),
image
3. As userB, click on the plus actions button in the compose message and select "request money"
4. Once requested have userA and userB click on the expense preview to view the expense report.
5. Confirm userA, the admin sees the subheader, `From <userB full name> in <workspace name>`
image
6. Confirm userB, the member sees the subheader `From <workspace name>`
image
7. Click on the subheader and confirm you are navigated up one level to parent.

### Expense chats children
1. Continuation from the previous section, click on a specific expense request.
2. Confirm both admin and member sees `From <Expense parent report name> in <workspace name>`. The Expense parent report name can be seen in the LHN of the expense report.
3. Create any number of threads or nested threads under the expense report and confirm the subheader for them all is `From <Expense parent report name> in <workspace name>`.
image
4. Click on the subheader and confirm you are always navigated up one level to parent.

- [x] Verify that no errors appear in the JS console

### Offline tests

None, haven't made any API changes so we're not testing optimistic/offline steps for these flows for now.

### QA Steps

Same as Web-tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

Without workspace (IOU report)
image

Without workspace (IOU report child)
image

With workspace (expense report)
image

With workspace (expense report child)
image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

Without workspace (IOU report)
image

Without workspace (IOU report child)
image

With workspace (expense report)
image

With workspace (expense report child)
image

</details>

<details>
<summary>Desktop</summary>

Without workspace (IOU report)
image

Without workspace (IOU report child)
image

With workspace (expense report)
image

With workspace (expense report child)
image

</details>

<details>
<summary>iOS</summary>

Without workspace (IOU report)
image

Without workspace (IOU report child)
image

With workspace (expense report)
image

With workspace (expense report child)
image

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/components/AvatarWithDisplayName.js
Comment (venture1981) [inline src/components/AvatarWithDisplayName.js:20]: seems like this line was not passed to lint. check the syntax again.

**Review by venture1981 — COMMENTED** [review]
@chiragsalian 
check the link error.

File: src/components/AvatarWithDisplayName.js
Comment (chiragsalian) [reply src/components/AvatarWithDisplayName.js:20]: PR is in a draft state and not ready to review just yet 🙂

File: src/components/MoneyRequestHeader.js
Comment (chiragsalian) [inline src/components/MoneyRequestHeader.js:97]: @luacmartins, can you check if this change and the one below where i remove parentReport from HeaderWithBackButton breaks anything with single expense/IOU view?

also cc @grgia for awareness as this code change affects report headers since we're providing source report instead of parent report to HeaderWithBackButton.

Comment (luacmartins) [reply src/components/MoneyRequestHeader.js:97]: Some of [these checks](https://github.com/Expensify/App/blob/38b095fdcda09ba3dabfba1da8052441a64ad93d/src/components/AvatarWithDisplayName.js#L57C1-L59) depend on the report type, which can be different with your changes. Previously, the report type of the parentReport on a singleTransactionView was `iou` or `expense` and now it'll always be `chat`.

The only difference that I noticed with your changes is that the Workspace subscript avatar is missing in the single transaction view. There might be others given the type checks we do in the code I linked above.

**Before**
expense-transaction

**After**
Screenshot 2023-06-21 at 1 09 39 PM

File: src/libs/ReportUtils.js
Comment (luacmartins) [inline src/libs/ReportUtils.js:532]: I think we could rename this to avoid confusion with `isChatThread`. Maybe `hasParentReport`?

**Review by luacmartins — CHANGES_REQUESTED** [review]
Looking good. Left a couple of comments

File: src/components/MoneyRequestHeader.js
Comment (chiragsalian) [reply src/components/MoneyRequestHeader.js:97]: Hmm i think that geogia is changing the avatar anyway. @grgia do you know what the final avatar for this single expense request is supposed to be. Is it supposed to be with a workspace subscript? If so i think i can easily add it in using the current report instead of having the parent report.

**Comment by chiragsalian** [comment]
> Should we consider and fix tooltip in all types of report headers in this PR?

Nope, this PR is just focusing on report subheaders and navigating to parent. Unless there is some regression or change that would negatively affect this [PR](https://github.com/Expensify/App/pull/20512). The PR https://github.com/Expensify/App/pull/20512 is focused on avatars and i think Georgia is making another PR to focus on headers. So here I don't plan on effecting those and just keeping my focus to subheaders. 

My code does change avatar a bit as mentioned by carlos over [here](https://github.com/Expensify/App/pull/21092#discussion_r1237528640) but I'm discussing over there with Georgia to confirm if its an issue or not. That might be the only spot i could make an avatar change or after we merge this PR Georgia could make the necessary change too.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js:1108]: We already have a `getChatRoomSubtitle` see [here](https://github.com/Expensify/App/blob/f10878fd12b5b47861a91d8684e9d76eebe0fee7/src/components/AvatarWithDisplayName.js#L57). This once is a bit different since it gives us the text for parent navigation and we use it few lines below [here](https://github.com/Expensify/App/blob/f10878fd12b5b47861a91d8684e9d76eebe0fee7/src/components/AvatarWithDisplayName.js#L63).

I agree it returns just a string and not a link. I'll rename this to `getParentNavigationText`

File: src/components/AvatarWithDisplayName.js
Comment (chiragsalian) [reply src/components/AvatarWithDisplayName.js:97]: I'll keep this as is for now since it's mostly the same logic for `subtitle` few lines below. I think if we ever change this to useCallback then we should change it for both. Let me know if you feel strongly about changing this now.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js:532]: i personally like `isThread` cause it's simple. `hasParentReport` feels a little misleading to me cause we check for parentReportID and parentReportActionID. So then i would have to name it `hasParentReportIDAndActionID`, which feels unnecessary to me because that's what a "thread" is.

isChatThread - just checks if isThread and type of report is chat. I can easily see us expanding to do isIOUThread etc in future if needed too 🤷‍♂️

Let me know if you still feel differently about this.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js]: updated

File: src/libs/actions/Task.js
Comment (chiragsalian) [inline src/libs/actions/Task.js:60]: i noticed tasks do not save reportActionID optimistically. We need this so that functionality is smooth so i added it here.

**Comment by chiragsalian** [comment]
I made some minor improvements and this is ready for review again.

**Review by luacmartins — DISMISSED** [review]
LGTM. Just pending this discussion https://github.com/Expensify/App/pull/21092#discussion_r1237603909

**Comment by chiragsalian** [comment]
friendly bump for reviews @0xmiroslav / @grgia since this is nice to have for SNH.

**Review by 0xmiroslav — COMMENTED** [review]
Still in progress of testing. Will update soon

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js]: just updated the code to address it 🙂

File: src/components/HeaderWithBackButton.js
Comment (chiragsalian) [reply src/components/HeaderWithBackButton.js:191]: For sure, i would check everything besides avatar because Georgia is changing it up in a follow up issue. And even if she does not address it i can address it in a follow up so i wouldn't want to block us on the avatar alone.

File: src/components/AvatarWithDisplayName.js
Comment (chiragsalian) [reply src/components/AvatarWithDisplayName.js:95]: Can you share steps to get the AnonymousReportFooter? I don't know it off the top of my head and if you have the steps it would save me some time to check this 🙂

Comment (chiragsalian) [reply src/components/AvatarWithDisplayName.js:95]: Cool, i have an errand for the next 90 minutes, so once I'm back I'll check your comments. Meanwhile, if you find any other issues let me know 👍

Comment (chiragsalian) [reply src/components/AvatarWithDisplayName.js:95]: tested this and yes having it in AnonymousReportFooter looks fine to me for now.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js:1061]: I tried logging into a gmail test account and my Expensify account and it seemed to work fine for me. I could not reproduce this same error. Either way, i added a safety check for this so it should not throw an error. Feel free to retest.

**Comment by chiragsalian** [comment]
Updated code to address most recent concerns.

**Review by luacmartins — DISMISSED** [review]

**Review by 0xmiroslav — APPROVED** [review]
Tested all possible cases as stated in https://github.com/Expensify/App/pull/20512#pullrequestreview-1484085317

**Comment by chiragsalian** [comment]
Since SNH began we're pausing on merging this PR for now. We'll merge it once feature freeze is lifted i.e., on 29th.

**Review by grgia — COMMENTED** [review]
oops I never submitted these comments

File: src/libs/ReportUtils.js
Comment (luacmartins) [reply src/libs/ReportUtils.js]: haha https://github.com/Expensify/App/pull/21092#discussion_r1237374145 :marshmallowduet:

**Comment by luacmartins** [comment]
I'm gonna put a hold for the feature freeze so we don't accidentally merge this PR considering it touches several files.

**Comment by luacmartins** [comment]
Removed the feature freeze hold. @chiragsalian it looks like @grgia left a few comments.

**Comment by luacmartins** [comment]
@chiragsalian we have conflicts too!

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js:753]: Maybe but this is just a method name refactor. The functionality is the same as before so I'm not updating the logic here since its outside of the scope of this PR which already has complex test steps.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js:1049]: Similar comment as previous. I really don't want to change existing functionality in this PR since it's already large to test.

File: src/libs/ReportUtils.js
Comment (chiragsalian) [reply src/libs/ReportUtils.js]: Ah yup, addressed this a while ago.

**Comment by chiragsalian** [comment]
oops i missed georgia's comments earlier. I was just thinking as soon as feature freeze was done we would merge this. Either way, i replied to her comments and updated the PR to address merge conflicts.

Feel free to review/merge.

**Comment by grgia** [comment]
Pulled and testing a few cases locally!

**Review by grgia — APPROVED** [review]
changes LGTM and tests well!

**Review by thienlnam — APPROVED** [review]
Approving and sending since we have all approvals and quite a few issues are holding on this

**Pull request merged**
